File: breakpilot-vscode/src/security/ConsentManager.ts
import * as vscode from "vscode";
import { BridgeClient } from "../bridge/BridgeClient";
import { BridgeMessage, MessageTypes } from "../bridge/MessageProtocol";
import { ConfirmationChoice, showConfirmationDialog } from "../ui/ConfirmationDialog";

type SafeInspectionsMode = "ask_once_per_project" | "always_ask" | "always_allow_trusted_projects";
type DebugControlsMode = "ask_once_per_session" | "always_ask";

const SAFE_ACTIONS_KEY = "breakpilot.allowedSafeInspectionActions";

export class ConsentManager {
  private readonly sessionAllowedActions = new Map<string, Set<string>>();

  constructor(private readonly context: vscode.ExtensionContext) {}

  async handleConfirmation(bridge: BridgeClient, message: BridgeMessage) {
    const autoScope = this.automaticRememberScope(message);
    if (autoScope) {
      this.sendAllow(bridge, message, autoScope);
      return;
    }

    const choice = await showConfirmationDialog(message);
    if (!choice.allowed) {
      this.sendReject(bridge, message);
      return;
    }

    await this.remember(message, choice.rememberScope);
    this.sendAllow(bridge, message, choice.rememberScope);
  }

  clearSession(ideSessionId: string) {
    this.sessionAllowedActions.delete(ideSessionId);
  }

  async resetDecisions() {
    this.sessionAllowedActions.clear();
    await this.context.workspaceState.update(SAFE_ACTIONS_KEY, []);
  }

  private automaticRememberScope(message: BridgeMessage): string | null {
    const action = message.action;
    if (!action) return null;
    const risk = message.riskLevel ?? "control";
    if (risk === "safe") return this.safeInspectionScope(action);
    if (risk === "control") return this.debugControlScope(message.ideSessionId, action);
    if (risk === "high") return this.highRiskAllowlistScope(message);
    return null;
  }

  private safeInspectionScope(action: string): string | null {
    const mode = this.config<SafeInspectionsMode>("safeInspectionsMode", "ask_once_per_project");
    if (mode === "always_ask") return null;
    if (mode === "always_allow_trusted_projects") return this.trustedProject() ? "project" : null;
    return this.safeInspectionAllowed(action) ? "project" : null;
  }

  private debugControlScope(ideSessionId: string | undefined, action: string): string | null {
    const mode = this.config<DebugControlsMode>("debugControlsMode", "ask_once_per_session");
    if (mode === "always_ask" || !ideSessionId) return null;
    return this.sessionAllowedActions.get(ideSessionId)?.has(action) ? "session" : null;
  }

  private highRiskAllowlistScope(message: BridgeMessage): string | null {
    if (!this.config<boolean>("allowPersistentHighRiskApprovals", false)) return null;
    if (this.config<boolean>("allowlistTrustedProjectsOnly", true) && !this.trustedProject()) return null;

    const action = message.action ?? "";
    if (this.config<string[]>("allowedActions", ["readonly_evaluate"]).includes(action)) return "project";

    const expression = this.expressionForAllowlist(message);
    return this.config<string[]>("allowedExpressionPatterns", []).some((pattern) => this.patternMatches(pattern, expression))
      ? "project"
      : null;
  }

  private async remember(message: BridgeMessage, scope: string) {
    const action = message.action;
    if (!action) return;
    if (scope === "project" && message.riskLevel === "safe") {
      const existing = new Set(this.context.workspaceState.get<string[]>(SAFE_ACTIONS_KEY, []));
      existing.add(action);
      await this.context.workspaceState.update(SAFE_ACTIONS_KEY, [...existing]);
      return;
    }
    if (scope === "session") {
      const key = message.ideSessionId;
      if (!key) return;
      const allowed = this.sessionAllowedActions.get(key) ?? new Set<string>();
      allowed.add(action);
      this.sessionAllowedActions.set(key, allowed);
    }
  }

  private safeInspectionAllowed(action: string): boolean {
    return this.context.workspaceState.get<string[]>(SAFE_ACTIONS_KEY, []).includes(action);
  }

  private trustedProject(): boolean {
    return this.config<boolean>("trustedProject", false);
  }

  private expressionForAllowlist(message: BridgeMessage): string {
    const payloadExpression = message.payload?.expression;
    if (typeof payloadExpression === "string") return payloadExpression;
    return message.expressionPreview ?? "";
  }

  private patternMatches(pattern: string, value: string): boolean {
    if (!pattern || !value) return false;
    try {
      return new RegExp(pattern).test(value);
    } catch {
      return value.includes(pattern);
    }
  }

  private config<T>(key: string, fallback: T): T {
    return vscode.workspace.getConfiguration("breakpilot").get<T>(key, fallback);
  }

  private sendAllow(bridge: BridgeClient, message: BridgeMessage, rememberScope: string) {
    bridge.send({
      type: MessageTypes.UserConfirmContinue,
      confirmationId: message.confirmationId,
      sessionId: message.sessionId,
      ideSessionId: message.ideSessionId,
      pauseEpoch: message.expectedPauseEpoch,
      originRequestId: message.originRequestId,
      action: message.action ?? "allow",
      rememberScope
    });
  }

  private sendReject(bridge: BridgeClient, message: BridgeMessage) {
    bridge.send({
      type: MessageTypes.UserRejectContinue,
      confirmationId: message.confirmationId,
      sessionId: message.sessionId,
      ideSessionId: message.ideSessionId,
      pauseEpoch: message.expectedPauseEpoch,
      originRequestId: message.originRequestId
    });
  }
}

export type { ConfirmationChoice };

File: breakpilot-idea/src/main/kotlin/security/ConsentManager.kt
package security

import bridge.BridgeClient
import bridge.BridgeMessage
import bridge.MessageTypes
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.service
import com.intellij.openapi.project.Project
import settings.BreakPilotProjectConsentState
import settings.BreakPilotSettingsState
import settings.DebugControlsMode
import settings.SafeInspectionsMode
import ui.showBreakpointConfirmation

@Service(Service.Level.PROJECT)
class ConsentManager(private val project: Project) {
    // Session-level consent is intentionally memory-only. A fresh debug session
    // should ask again because continue/step/stop affect the user's live flow.
    private val sessionAllowedActions = mutableMapOf<String, MutableSet<String>>()

    fun handleConfirmation(bridge: BridgeClient, message: BridgeMessage) {
        // Prefer an existing decision before showing UI. The server still sees a
        // normal confirmation response, so audit/control flow stays unchanged.
        val autoScope = automaticRememberScope(message)
        if (autoScope != null) {
            sendAllow(bridge, message, autoScope)
            return
        }

        val choice = showBreakpointConfirmation(message)
        if (!choice.allowed) {
            sendReject(bridge, message)
            return
        }

        remember(message, choice.rememberScope)
        sendAllow(bridge, message, choice.rememberScope)
    }

    fun clearSession(ideSessionId: String) {
        sessionAllowedActions.remove(ideSessionId)
    }

    fun resetSessionDecisions() {
        sessionAllowedActions.clear()
    }

    private fun automaticRememberScope(message: BridgeMessage): String? {
        // Risk level is supplied by the runtime policy layer; IDEA only decides
        // whether the user has already granted the permitted remember scope.
        val risk = message.riskLevel ?: "control"
        val action = message.action ?: return null
        return when (risk) {
            "safe" -> safeInspectionScope(action)
            "control" -> debugControlScope(message.ideSessionId, action)
            "high" -> highRiskAllowlistScope(message)
            else -> null
        }
    }

    private fun safeInspectionScope(action: String): String? {
        val settings = service<BreakPilotSettingsState>().state
        val projectConsent = project.service<BreakPilotProjectConsentState>()
        // Safe inspection can be remembered at project scope because it reads
        // paused state without resuming or mutating the debuggee.
        return when (SafeInspectionsMode.fromId(settings.safeInspectionsMode)) {
            SafeInspectionsMode.AlwaysAsk -> null
            SafeInspectionsMode.AlwaysAllowTrustedProjects ->
                if (projectConsent.state.trustedProject) "project" else null
            SafeInspectionsMode.AskOncePerProject ->
                if (projectConsent.isSafeInspectionAllowed(action)) "project" else null
        }
    }

    private fun debugControlScope(ideSessionId: String?, action: String): String? {
        val settings = service<BreakPilotSettingsState>().state
        if (DebugControlsMode.fromId(settings.debugControlsMode) == DebugControlsMode.AlwaysAsk) return null
        val key = ideSessionId ?: return null
        return if (sessionAllowedActions[key]?.contains(action) == true) "session" else null
    }

    private fun highRiskAllowlistScope(message: BridgeMessage): String? {
        val settings = service<BreakPilotSettingsState>().state
        if (!settings.allowPersistentHighRiskApprovals) return null
        val projectConsent = project.service<BreakPilotProjectConsentState>().state
        if (settings.allowlistTrustedProjectsOnly && !projectConsent.trustedProject) return null

        // Persistent high-risk approval is never broad by default. It requires
        // the advanced setting plus a narrow action or expression allowlist hit.
        val action = message.action.orEmpty()
        if (settings.allowedActions.contains(action)) return "project"

        val expression = (message.payload["expression"] as? String) ?: message.expressionPreview.orEmpty()
        return if (settings.allowedExpressionPatterns.any { patternMatches(it, expression) }) "project" else null
    }

    private fun patternMatches(pattern: String, value: String): Boolean {
        if (pattern.isBlank() || value.isBlank()) return false
        return try {
            Regex(pattern).containsMatchIn(value)
        } catch (_: Throwable) {
            // Invalid regex entries degrade to simple substring matching so a
            // typo does not break the settings page or confirmation flow.
            value.contains(pattern)
        }
    }

    private fun remember(message: BridgeMessage, scope: String) {
        val action = message.action ?: return
        when (scope) {
            "project" -> if (message.riskLevel == "safe") {
                project.service<BreakPilotProjectConsentState>().allowSafeInspection(action)
            }
            "session" -> {
                val key = message.ideSessionId ?: return
                sessionAllowedActions.getOrPut(key) { mutableSetOf() }.add(action)
            }
        }
    }

    private fun sendAllow(bridge: BridgeClient, message: BridgeMessage, rememberScope: String) {
        bridge.send(
            BridgeMessage(
                type = MessageTypes.UserConfirmContinue,
                confirmationId = message.confirmationId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch,
                action = message.action ?: "allow",
                rememberScope = rememberScope
            )
        )
    }

    private fun sendReject(bridge: BridgeClient, message: BridgeMessage) {
        bridge.send(
            BridgeMessage(
                type = MessageTypes.UserRejectContinue,
                confirmationId = message.confirmationId,
                sessionId = message.sessionId,
                ideSessionId = message.ideSessionId,
                originRequestId = message.originRequestId,
                pauseEpoch = message.expectedPauseEpoch
            )
        )
    }
}
